internal/mesh/internal/types/destinations.go

```diff
@@ -93,11 +109,78 @@ func ValidateDestinations(res *pbresource.Resource) error {
 			merr = multierror.Append(merr, refErr)
 		}
 
-		// TODO(v2): validate port name using catalog validator
-		// TODO(v2): validate ListenAddr
+		if portErr := catalog.ValidatePortName(dest.DestinationPort); portErr != nil {
+			merr = multierror.Append(merr, wrapDestErr(resource.ErrInvalidField{
+				Name:    "destination_port",
+				Wrapped: portErr,
+			}))
+		}
+
+		if dest.GetDatacenter() != "" {
+			merr = multierror.Append(merr, wrapDestErr(resource.ErrInvalidField{
+				Name:    "datacenter",
+				Wrapped: resource.ErrUnsupported,
+			}))
+		}
+
+		if listenAddrErr := validateListenAddr(dest); listenAddrErr != nil {
+			merr = multierror.Append(merr, wrapDestErr(listenAddrErr))
+		}
+	}
+
+	return merr
+}
+
+func validateListenAddr(dest *pbmesh.Destination) error {
+	var merr error
+
+	if dest.GetListenAddr() == nil {
+		return multierror.Append(merr, resource.ErrInvalidFields{
+			Names:   []string{"ip_port", "unix"},
+			Wrapped: resource.ErrMissingOneOf,
+		})
+	}
+
+	switch dest.GetListenAddr().(type) {
+	case *pbmesh.Destination_IpPort:
+		listenAddr := dest.GetListenAddr().(*pbmesh.Destination_IpPort).IpPort
+
+		if ipPortErr := validateIPPort(listenAddr); ipPortErr != nil {
+			merr = multierror.Append(merr, resource.ErrInvalidField{
+				Name:    "ip_port",
+				Wrapped: ipPortErr,
+			})
+		}
+	case *pbmesh.Destination_Unix:
+		listenAddr := dest.GetListenAddr().(*pbmesh.Destination_Unix).Unix
+
+		if !catalog.IsValidUnixSocketPath(listenAddr.GetPath()) {
+			merr = multierror.Append(merr, resource.ErrInvalidField{
+				Name: "unix",
+				Wrapped: resource.ErrInvalidField{
+					Name:    "path",
+					Wrapped: errInvalidUnixSocketPath,
+				},
+			})
+		}
+	}
```

Comment:
Nit: I think this is equivalent
```suggestion
	switch listenAddr := dest.GetListenAddr().(type) {
	case *pbmesh.Destination_IpPort:
		if ipPortErr := validateIPPort(listenAddr); ipPortErr != nil {
			merr = multierror.Append(merr, resource.ErrInvalidField{
				Name:    "ip_port",
				Wrapped: ipPortErr,
			})
		}
	case *pbmesh.Destination_Unix:
		if !catalog.IsValidUnixSocketPath(listenAddr.GetPath()) {
			merr = multierror.Append(merr, resource.ErrInvalidField{
				Name: "unix",
				Wrapped: resource.ErrInvalidField{
					Name:    "path",
					Wrapped: errInvalidUnixSocketPath,
				},
			})
		}
	}
```